File: Spring Boot version/src/main/java/Thalassa/Controllers/UserController.java
package Thalassa.Controllers;

import Thalassa.Configuration.Constants;
import Thalassa.Services.UserService;
import Thalassa.Services.CryptographyService;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;
import java.util.HashMap;

@RestController
public class UserController {

    private final UserService userService;

    public UserController(UserService userService) {
        this.userService = userService;
    }

    @GetMapping("/connectionTest")
    public boolean connectionTest() {
        return true;
    }

    // Checks if the e-mail provided by the user during registration already exists and returns boolean
    @PostMapping("/emailExists")
    public boolean emailExists(@RequestBody String email) {
        return userService.findByEmail(email) != null;
    }

    // Generates user’s public and private keys
    @PostMapping(path = "/generateKeys", produces = MediaType.APPLICATION_JSON_VALUE)
    public ObjectNode generateKeys() {
        String[] keys;
        // Generate keys and check if public key is already used
        do keys = CryptographyService.generateKeys();
        while (userService.findByPublicKey(keys[0]) != null);

        ObjectNode json = new ObjectMapper().createObjectNode();
        json.put("publicKey", keys[0]);
        json.put("privateKey", keys[1]);

        return json;
    }

    // Validates registration form before final registration
    @PostMapping(path = "/registerIsValid", consumes = MediaType.APPLICATION_JSON_VALUE)
    public boolean registerIsValid(@RequestBody ObjectNode json) throws JsonProcessingException {
        return userService.validateRegister(json);
    }

    // Returns localized messages
    @GetMapping(path = "/localization", produces = MediaType.APPLICATION_JSON_VALUE)
    public HashMap<String, HashMap<String, String>> localization() {
        return Constants.LOCALIZATION;
    }
}

File: ASP.NET Core version/Controllers/UserController.cs
using System.Collections.Generic;
using System.IO;
using System.Text;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using Newtonsoft.Json.Linq;
using Thalassa.Services;

namespace Thalassa.Controllers
{
    [ApiController, Route("[controller]")]
    public class UserController : ControllerBase
    {
        private readonly IUserService _userService;

        public UserController(IUserService userService)
        {
            _userService = userService;
        }

        // Checks if the e-mail provided by the user during registration already exists and returns boolean
        [HttpPost("EmailExists")]
        public bool EmailExists()
        {
            using var reader = new StreamReader(Request.Body, Encoding.UTF8);
            return _userService.FindByEmail(reader.ReadToEnd()) != null;
        }

        // Generates user’s public and private keys
        [HttpPost("GenerateKeys")]
        public JObject GenerateKeys()
        {
            string[] keys;
            // Generate keys and check if public key is already used
            do keys = CryptographyService.GenerateKeys();
            while (_userService.FindByPublicKey(keys[0]) != null);

            return new JObject {["publicKey"] = keys[0], ["privateKey"] = keys[1]};
        }

        // Validates registration form before final registration
        [HttpPost("RegisterIsValid")]
        public bool RegisterIsValid([FromBody] JObject json)
        {
            return _userService.ValidateRegister(json);
        }
        
        [HttpPost("Register")]
        public JObject Register([FromBody] JObject json)
        {
            var result = _userService.Register(json);
            if (result.ContainsKey("token"))
                HttpContext.Session.SetString("loggedInPrivateKey", json["fields"]["privateKey"].Value<string>());

            return result;
        }

        [HttpPost("Login")]
        public JObject Login([FromBody] JObject json)
        {
            var result = _userService.Login(json);
            if (!result.ContainsKey("invalidFields") && !result.ContainsKey("noUser"))
                HttpContext.Session.SetString("loggedInPrivateKey", json["fields"]["privateKey"].Value<string>());

            return result;
        }

        // Returns localized messages
        [HttpGet("Localization")]
        public Dictionary<string, Dictionary<string, string>> Localization()
        {
            return Constants.Localization;
        }
        
        [Route("Logout")]
        public void Logout()
        {
            foreach (var cookie in Request.Cookies.Keys)
                Response.Cookies.Delete(cookie);
            HttpContext.Session.Clear();
        }
    }
}